File: backend/schemas/backtest.py
from datetime import datetime, date
from pydantic import BaseModel


class BacktestIn(BaseModel):
    symbol: str
    start_date: date
    end_date: date
    initial_balance: float = 100000.0
    portfolio_id: int = 1
    signal_every_n_bars: int = 5
    max_position_pct: float = 0.20


class BacktestOut(BaseModel):
    id: int
    portfolio_id: int
    symbol: str
    start_date: date
    end_date: date
    initial_balance: float
    final_balance: float | None
    total_trades: int | None
    winning_trades: int | None
    max_drawdown: float | None
    sharpe_ratio: float | None
    total_return: float | None
    win_rate: float | None
    status: str
    error_message: str | None
    created_at: datetime
    completed_at: datetime | None

    model_config = {"from_attributes": True}


class BacktestResultOut(BacktestOut):
    results_json: str | None = None

File: frontend/src/api/backtest.ts
import client from './client'

export interface Backtest {
  id: number
  portfolio_id: number
  symbol: string
  start_date: string
  end_date: string
  initial_balance: number
  final_balance: number | null
  total_trades: number | null
  winning_trades: number | null
  max_drawdown: number | null
  sharpe_ratio: number | null
  total_return: number | null
  win_rate: number | null
  status: string
  error_message: string | null
  created_at: string
  completed_at: string | null
  results_json?: string | null
}

export const runBacktest = (params: {
  symbol: string
  start_date: string
  end_date: string
  initial_balance?: number
  signal_every_n_bars?: number
  max_position_pct?: number
}) => client.post<Backtest>('/backtest/run', params).then((r) => r.data)

export const getBacktests = () => client.get<Backtest[]>('/backtest').then((r) => r.data)

export const getBacktest = (id: number) => client.get<Backtest>(`/backtest/${id}`).then((r) => r.data)

export const deleteBacktest = (id: number) => client.delete(`/backtest/${id}`).then((r) => r.data)
